src/credentials/LastKnownGoodTime.cpp

```diff
@@ -0,0 +1,251 @@
+    // Verify that the passed value is not earlier than the passed NotBefore time.
+    VerifyOrExit(lastKnownGoodChipEpochTime >= notBefore, err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Passed value is valid.  Capture it and write back to persistence.
+    //
+    // Note that we are purposefully overwriting any previous last known
+    // good time that may have been stored as part of a fail-safe context.
+    // This is intentional: we don't promise not to change last known good
+    // time during the fail-safe timer.  For instance, if the platform has a
+    // new, better time source, it is legal to capture it.  If we do, we should
+    // both overwrite last known good time and discard the previous value stored
+    // for fail safe recovery, as together these comprise a transaction.  By
+    // overwriting both, we are fully superseding that transaction with our
+    // own, which does not to have a revert feature.
+    SuccessOrExit(err = StoreLastKnownGoodChipEpochTime(lastKnownGoodChipEpochTime));
+    mLastKnownGoodChipEpochTime.SetValue(lastKnownGoodChipEpochTime);
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(TimeService, "Failed to update Last Known Good Time: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+    else
+    {
+        LogTime("Updating Last Known Good Time to ", lastKnownGoodChipEpochTime);
+    }
+    return err;
+}
+
+CHIP_ERROR LastKnownGoodTime::UpdateLastKnownGoodChipEpochTime(System::Clock::Seconds32 lastKnownGoodChipEpochTime)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(mLastKnownGoodChipEpochTime.HasValue(), err = CHIP_ERROR_INCORRECT_STATE);
+    LogTime("Last Known Good Time: ", mLastKnownGoodChipEpochTime.Value());
+    LogTime("New proposed Last Known Good Time: ", lastKnownGoodChipEpochTime);
+    if (lastKnownGoodChipEpochTime > mLastKnownGoodChipEpochTime.Value())
+    {
+        LogTime("Current Last Known Good time retained in fail-safe context, updating to ", lastKnownGoodChipEpochTime);
+        // We have a later timestamp.  Advance last known good time and store
+        // the failsafe value.
+        SuccessOrExit(
+            err = StoreLastKnownGoodChipEpochTime(lastKnownGoodChipEpochTime, MakeOptional(mLastKnownGoodChipEpochTime.Value())));
+        mLastKnownGoodChipEpochTime.SetValue(lastKnownGoodChipEpochTime);
+    }
+    else
+    {
+        ChipLogProgress(TimeService, "Retaining current Last Known Good Time");
+        // Our timestamp is not later.  Retain the existing last known good time
+        // and discard any fail-safe value in persistence.
+        SuccessOrExit(err = StoreLastKnownGoodChipEpochTime(mLastKnownGoodChipEpochTime.Value()));
+    }
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(TimeService, "Failed to persist Last Known Good Time: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+    return err;
+}
+
+CHIP_ERROR LastKnownGoodTime::RevertLastKnownGoodChipEpochTime()
```

Comment:
So this is where we revert on fail-safe expiration... but where do we commit on CommissioningComplete?  It looks to me like we could start fail-safe window, add a cert, CommissioningComplete, then start another fail-safe window, let it expire, and we will revert the value from that cert we added....

Reply:
Ah, you are correct.  I'll fix that.

Reply:
Fixed.